src/site/content/en/blog/webassembly-memory-debugging/index.md

```diff
@@ -0,0 +1,607 @@
+> x.method();
+> x.delete();
+> ```
+
+Indeed, when we do that in JavaScript for our class:
+
+```ts/7/
+  // …
+
+  const result = opts.zx
+    ? module.zx_quantize(data.data, data.width, data.height, opts.dither)
+    : module.quantize(data.data, data.width, data.height, opts.maxNumColors, opts.dither);
+
+  module.free_result();
+  result.delete();
+  module.doLeakCheck();
+
+  return new ImageData(
+    new Uint8ClampedArray(result.view),
+    result.width,
+    result.height
+  );
+}
+```
+
+The leak goes away as expected.
+
+### Discovering more issues with sanitizers
+
+Building other Squoosh codecs with sanitizers reveals both similar as well as some new issues. For
+example, I've got this error in MozJPEG bindings:
+
+![Screenshot of a message "memory access out of bounds" coming from a `jpeg_start_compress` function](out-of-bounds.png)
+
+Here, it's not a leak, but us writing to a memory outside of the allocated boundaries 😱
+
+Digging into the code of MozJPEG, we find that the problem here is that `jpeg_mem_dest`—the
+function that we use to allocate a memory destination for JPEG—[reuses existing values of
+`outbuffer` and `outsize` when they're
+non-zero](https://github.com/mozilla/mozjpeg/blob/1d2320994dd0d293d39cfaea3d13060b60f32c45/jdatadst.c#L282-L288):
+
+```cpp/0
+if (*outbuffer == NULL || *outsize == 0) {
+  /* Allocate initial buffer */
+  dest->newbuffer = *outbuffer = (unsigned char *) malloc(OUTPUT_BUF_SIZE);
+  if (dest->newbuffer == NULL)
+    ERREXIT1(cinfo, JERR_OUT_OF_MEMORY, 10);
+  *outsize = OUTPUT_BUF_SIZE;
+}
+```
+
+However, we invoke it without initialising either of those variables, which means MozJPEG writes the
+result into a potentially random memory address that happened to be stored in those variables at the
+time of the call!
+
+```cpp/0-1
+uint8_t* output;
+unsigned long size;
+// …
+jpeg_mem_dest(&cinfo, &output, &size);
+```
+
+Zero-initialising both variables before the invocation solves this issue, and now the code reaches a
+memory leak check instead. Luckily, the check passes successfully, indicating that we don't have any
+leaks in this codec.
+
+## Issues with shared state
+
+…Or do we?
+
+We know that our codec bindings store some of the state as well as results in global static
+variables, and MozJPEG has some particularly complicated structures.
+
+```cpp
+uint8_t* last_result;
+struct jpeg_compress_struct cinfo;
+
+val encode(std::string image_in, int image_width, int image_height, MozJpegOptions opts) {
+  // …
+}
+```
+
+What if some of those get lazily initialized on the first run, and then improperly reused on future
+runs? Then a single call with a sanitizer would not report them as problematic.
+
+Let's try and process the image a couple of times by randomly clicking at different quality levels
+in the UI. Indeed, now we get the following report:
+
+![Screenshot of a message "Direct leak of 262144 bytes" coming from a `jpeg_finish_compress` function](leak2.png)
```

Comment:
Please use this formatting: https://web.dev/handbook/markup-media/#partial-screenshots